File: garden/seeds/java/app/src/main/java/com/WATER_ME_WITH_COMPANY/android/WATER_ME_WITH_APP_NAME/module/WATER_ME_WITH_CLASS_PREFIX_Module.java
package com.{{company_name}}.android.{{app_package_name_prefix}}.module;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import com.{{company_name}}.android.{{app_package_name_prefix}}.{{app_class_prefix}}App;
import com.{{company_name}}.android.{{app_package_name_prefix}}.data.AppSettings;
import com.lacronicus.easydatastorelib.DatastoreBuilder;

import com.squareup.leakcanary.RefWatcher;

import java.util.LinkedList;
import java.util.List;

import javax.inject.Named;
import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

/**
 * Dagger module to provide dependency injection
 */
@SuppressWarnings("unused")
@Module
public class {{app_class_prefix}}Module {
    private {{app_class_prefix}}App application;

    public {{app_class_prefix}}Module({{app_class_prefix}}App app) {
        application = app;
    }

    @Provides
    @Singleton
    {{app_class_prefix}}App providesApp() {
        return application;
    }

    @Provides
    Context providesContext() {
        return application;
    }

    @Provides
    RefWatcher providesRefWatcher({{app_class_prefix}}App app) {
        return app.getRefWatcher();
    }

    @Provides
    @Singleton
    SharedPreferences providesSharedPreferences(Context context) {
        return PreferenceManager.getDefaultSharedPreferences(context);
    }

    @Provides
    @Singleton
    AppSettings providesAppSettings(SharedPreferences prefs) {
        return new DatastoreBuilder(prefs).create(AppSettings.class);
    }
}

File: garden/seeds/kotlin/app/src/main/kotlin/com/WATER_ME_WITH_COMPANY/android/WATER_ME_WITH_APP_NAME/module/WATER_ME_WITH_CLASS_PREFIX_Module.kt
package com.{{company_name}}.android.{{app_package_name_prefix}}.module

import android.content.Context
import android.content.SharedPreferences
import android.preference.PreferenceManager
import com.{{company_name}}.android.{{app_package_name_prefix}}.BuildConfig
import com.{{company_name}}.android.{{app_package_name_prefix}}.{{app_class_prefix}}App
import com.{{company_name}}.android.{{app_package_name_prefix}}.{{app_class_prefix}}AppImpl
import com.{{company_name}}.android.{{app_package_name_prefix}}.data.AppSettings
import com.{{company_name}}.{{app_package_name_prefix}}.DataSource
import com.{{company_name}}.{{app_package_name_prefix}}.networkDataSource
import com.lacronicus.easydatastorelib.DatastoreBuilder
import dagger.Module
import dagger.Provides
import retrofit.RestAdapter
import javax.inject.Singleton

/**
 * Dagger module to provide dependency injection
 */
SuppressWarnings("UNUSED_PARAMETER")
Module
public class {{app_class_prefix}}Module(private val application: {{app_class_prefix}}App) {

    Provides
    Singleton
    fun providesApp(): {{app_class_prefix}}App {
        return application
    }

    Provides
    fun providesContext(): Context {
        return application
    }

    Provides
    Singleton
    fun providesSharedPreferences(context: Context): SharedPreferences {
        return PreferenceManager.getDefaultSharedPreferences(context)
    }

    Provides
    Singleton
    fun providesAppSettings(sharedPreferences: SharedPreferences): AppSettings {
        return DatastoreBuilder(sharedPreferences).create(javaClass<AppSettings>())
    }

    Provides
    Singleton
    fun providesDataSource() : DataSource {
        return networkDataSource {
            logging = BuildConfig.DEBUG
            endpoint = "http://myawesomeserver.com"
            networkInterceptors = listOf()
        }
    }
}
